web/lib/SuffixTree.ts

```diff
@@ -0,0 +1,55 @@
+export class SuffixTree {
+    root: Record<string, any>;
+    steps: Array<{ char: string; path: string[] }>;
+    nodeToParent: WeakMap<Record<string, any>, Record<string, any>>;
+
+    constructor() {
+        this.root = {};
+        this.steps = [];
+        this.nodeToParent = new WeakMap();
+        this.nodeToParent.set(this.root, null);
+    }
+
+    addSuffix(suffix: string) {
+        let node = this.root;
+        for (const char of suffix) {
+            if (!node[char]) {
+                node[char] = {};
+                this.steps.push({ char, path: this.getPath(node) });
+                this.nodeToParent.set(node[char], node);
+            }
+            node = node[char];
+        }
+        node.isEnd = true;
+    }
+
+    addString(str: string) {
+        console.log(`Adding string: ${str}`);
+        for (let i = 0; i < str.length; i++) {
+            this.addSuffix(str.slice(i));
+            console.log(`Added suffix: ${str.slice(i)}`);
+        }
+    }
+
+    getPath(node: Record<string, any>): string[] {
```

Comment:
some comments would be helpful here for a reader to understand what's going on. Do we traverse from the given node back to the root while building and returning a path?